src/deepsparse/pipeline.py

```diff
@@ -176,22 +174,113 @@ def __init__(
-            executor = kwargs.pop("executor")
+            postprocess_kwargs = {}
+
+        # split inputs into batches of size `self._batch_size`
+        batches = self.split_engine_inputs(engine_inputs, self._batch_size)
+
+        # submit to engine
+        futures = [
+            self.executor.submit(self.engine_forward, batch) for batch in batches
+        ]
+        wait(futures)
+
+        # join together the batches of size `self._batch_size`
+        engine_outputs = self.join_engine_outputs(
+            [future.result() for future in futures]
+        )
 
-        return (
-            executor.submit(self._run, *args, **kwargs)  # Non-Blocking call
-            if executor and not self.use_dynamic_batch()
-            else self._run(*args, **kwargs)  # Blocking call
+        pipeline_outputs = self.process_engine_outputs(
+            engine_outputs, **postprocess_kwargs
         )
+        if not isinstance(pipeline_outputs, self.output_schema):
+            raise ValueError(
+                f"Outputs of {self.__class__} must be instances of "
+                f"{self.output_schema} found output of type {type(pipeline_outputs)}"
+            )
+
+        return pipeline_outputs
+
+    def split_engine_inputs(
+        self, items: List[numpy.ndarray], batch_size: int
+    ) -> List[List[numpy.ndarray]]:
+        """
+        Splits each item into numpy arrays with the first dimension == `batch_size`.
+
+        For example, if `items` has three numpy arrays with the following
+        shapes: `[(4, 32, 32), (4, 64, 64), (4, 128, 128)]`
+
+        Then with `batch_size==4` the output would be:
+        ```
+        [[(4, 32, 32), (4, 64, 64), (4, 128, 128)]]
+        ```
+
+        Then with `batch_size==2` the output would be:
+        ```
+        [
+            [(2, 32, 32), (2, 64, 64), (2, 128, 128)],
+            [(2, 32, 32), (2, 64, 64), (2, 128, 128)],
+        ]
+        ```
+
+        Then with `batch_size==1` the output would be:
+        ```
+        [
+            [(1, 32, 32), (1, 64, 64), (1, 128, 128)],
+            [(1, 32, 32), (1, 64, 64), (1, 128, 128)],
+            [(1, 32, 32), (1, 64, 64), (1, 128, 128)],
+            [(1, 32, 32), (1, 64, 64), (1, 128, 128)],
+        ]
+        ```
+        """
+        assert all(isinstance(item, numpy.ndarray) for item in items)
+
+        # all items should have the same batch size
+        total_batch_size = items[0].shape[0]
+        assert all(item.shape[0] == total_batch_size for item in items)
+        if total_batch_size % batch_size != 0:
+            raise RuntimeError(
+                f"batch size of {total_batch_size} passed into pipeline "
+                f"is not divisible by model batch size of {batch_size}"
+            )
+
+        batches = []
+        for i_batch in range(total_batch_size // batch_size):
+            start = i_batch * batch_size
+            batches.append([item[start : start + batch_size] for item in items])
```

Comment:
does numpy need to write to new memory for this to get a contiguous array?

Reply:
No, it is copied:
```python
>>> q = np.random.rand(2, 5)
>>> q
array([[0.26492997, 0.57344731, 0.79898817, 0.42353296, 0.05981831],
       [0.17760227, 0.97477584, 0.74938403, 0.61339915, 0.23853161]])
>>> a, b = q[0:1], q[1:2]
>>> a
array([[0.26492997, 0.57344731, 0.79898817, 0.42353296, 0.05981831]])
>>> b
array([[0.17760227, 0.97477584, 0.74938403, 0.61339915, 0.23853161]])
>>> a *= 2
>>> a
array([[0.52985993, 1.14689462, 1.59797635, 0.84706591, 0.11963661]])
>>> q
array([[0.52985993, 1.14689462, 1.59797635, 0.84706591, 0.11963661],
       [0.17760227, 0.97477584, 0.74938403, 0.61339915, 0.23853161]])
```